### PR TITLE
Add logic to enable chunked MLP during training

### PR DESCRIPTION
**Pull request by pengdurice**
# What does this PR do ?

This PR enables chunked MLP during training phases (on top of prefilling of inference phase)
## related issue:
https://github.com/NVIDIA/Megatron-LM/issues/3644
## Tests:
Added unit tests and passed.
Applied to a multi node GLM4.7 model training and unblocked OOM with context window > 30K
## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [X] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [X] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [X] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by pengdurice** [comment]
@Phlip79 , Hi, would you mind reviewing this PR? thanks!

**Review by Phlip79 — CHANGES_REQUESTED** [review]
Could you update the `test()` helper to pass `mlp_chunks_for_training` into `TransformerConfig` instead of hardcoding it to 4, so the loop actually verifies that chunked and non-chunked MLP produce identical outputs?

**Comment by pengdurice** [comment]
> Could you update the `test()` helper to pass `mlp_chunks_for_training` into `TransformerConfig` instead of hardcoding it to 4, so the loop actually verifies that chunked and non-chunked MLP produce identical outputs?

Hi @Phlip79 , thank you for the review, fixed it;-)

File: megatron/core/transformer/transformer_layer.py [hunk 1]
```diff
@@ -742,6 +742,11 @@ def _forward_mlp(
             and not isinstance(self.mlp, IdentityOp)
             and not self.config.transformer_impl == "inference_optimized"
         )
+        should_chunk_mlp_for_training = (
+            self.config.mlp_chunks_for_training > 1
+            and inference_context is None
```
Comment (Phlip79) [inline megatron/core/transformer/transformer_layer.py:778]: Does this need `and self.training`?

File: tests/unit_tests/transformer/test_transformer_layer.py [hunk 2]
```diff
@@ -130,7 +133,25 @@ def test(
...
+                    input_hidden_states,
+                    inference_context,
+                )
+                assert hidden_states.shape[0] == sequence_length
+                assert hidden_states.shape[1] == micro_batch_size
+                assert hidden_states.shape[2] == hidden_size
+
+                outputs[mlp_chunks_for_training] = (hidden_states, context)
+
+            assert torch.equal(outputs[1][0], outputs[4][0])
```
Comment (Phlip79) [inline tests/unit_tests/transformer/test_transformer_layer.py:131]: Each call to `test()` creates a new `TransformerLayer` with fresh random weights. I think the better approach is to create it before the call to `test()` and re-use it.

**Review by Phlip79 — CHANGES_REQUESTED** [review]

File: megatron/core/transformer/transformer_layer.py (hunk 1 above)
Comment (pengdurice) [reply megatron/core/transformer/transformer_layer.py:778]: I thought inference_context and self.training was mutually exclusive, yes, sure, let me add it.

File: tests/unit_tests/transformer/test_transformer_layer.py (hunk 2 above)
Comment (pengdurice) [reply tests/unit_tests/transformer/test_transformer_layer.py:131]: Thanks! I brought the code from test outside and then directly call the created transformer layer. I also locally tested if `transformer_config.mlp_chunks_for_training = mlp_chunks_for_training` works by printing the num_chunk (not shown here.) and verified it works.  At the same time, I changed inputs from torch.ones to torch.randn with a fixed seed to further verify that it works. LMK if you have any more questions, thanks!

**Review by Phlip79 — APPROVED** [review]

**Comment by pengdurice** [comment]
@Phlip79  Hi, thanks again for reviewing this PR. I’ve fixed a lint issue—would appreciate another set of eyes when you have time!

**Comment by pengdurice** [comment]
@Phlip79 I saw several errors from the CICD run I downloaded the errors and checked and believe they fail on other places and probably due to that this PR is too far behind main. I rebased and fixed the lint errors again. Once you've got a chance, would you please trigger the CI/CD again? thank you!

**Comment by pengdurice** [comment]
@Phlip79 Hi, sorry for the back and forth! The lint check failed on the last CI test due to that I merged the main instead of rebasing on main, and I also tried to fix the lint errors causing more files to be changed which then trigger even more lint errors. The clean solution is to cherry pick only my commits (my 4 files' changes) and rebase on main. Exactly what I did now. 
I tested locally confirmed that the tests still passed and also the following passed:
Now the PR contains only the 4 file changes.  
> pre-commit run --files \
>   megatron/core/transformer/transformer_config.py \
>   megatron/core/transformer/transformer_layer.py \
>   tests/unit_tests/models/test_mamba_moe_model.py \
>   tests/unit_tests/transformer/test_transformer_layer.py
>   black....................................................................Passed
>   pylint...................................................................Passed
>   isort....................................................................Passed

Would you please trigger CI again? thank you so much!

**Comment by pengdurice** [comment]
> /ok to test [2cb97c5](https://github.com/NVIDIA/Megatron-LM/commit/2cb97c5bbeb2fd7f7beb13b8fd8e138e28886bd5)

@Phlip79 thank you for triggering the test. Would you mind take another look at it as most tests passed (other than approval workflow)? Thank you so much!

**Comment by Phlip79** [comment]
All good now. Need review groups to take a look now.

**Comment by pengdurice** [comment]
> All good now. Need review groups to take a look now.

Great, thank you, let me know if the team needs anything from me;-)

File: megatron/core/transformer/transformer_config.py [hunk 3]
```diff
@@ -969,6 +969,9 @@ class TransformerConfig(ModelParallelConfig):
     mlp_chunks_for_prefill: int = 1
     """The number of chunks along the sequence dimension to use for MLP computation
     during prefill."""
+    mlp_chunks_for_training: int = 1
+    """The number of chunks along the sequence dimension to use for MLP computation
+    during training."""
```
Comment (santhnm2) [inline megatron/core/transformer/transformer_config.py:974]: Rather than having a separate flag, perhaps we should just consolidate into a single `mlp_chunks` flag and then enable it either when we are in training mode (i.e., `inference_context` is `None`) or we are in prefill mode (`inference_context` is not `None` and `inference_context.is_decode_only() == False`). The only scenario where we would hypothetically need two separate flags is if we were doing RL and wanted the training to chunk MLP but the inference to not chunk MLP (or vice versa). This seems like a fairly contrived setting though. @jon-barker any thoughts here?

Comment (jon-barker) [reply megatron/core/transformer/transformer_config.py:974]: I have never thought about this in the RL context and don't really have an opinion how likely we are to want to do this. But it seems like maybe the proposed interface is cleaner anyway and leaves open the option of setting it separately for inference and training.

Comment (santhnm2) [reply megatron/core/transformer/transformer_config.py:974]: Sure, we can keep it as is for now. If we decide to consolidate later we can do that in a backwards-compatible way.

File: tests/unit_tests/transformer/test_transformer_layer.py [hunk 4]
```diff
@@ -64,73 +64,71 @@ def test_gpu_forward(self):
...
+                    hidden_states=input_hidden_states,
+                    attention_mask=attention_mask,
+                    inference_context=None,
+                )
+                assert hidden_states.shape[0] == sequence_length
+                assert hidden_states.shape[1] == micro_batch_size
+                assert hidden_states.shape[2] == hidden_size
+                outputs[mlp_chunks_for_training] = (hidden_states, context)
+
+            assert torch.equal(outputs[1][0], outputs[4][0])
```
Comment (santhnm2) [inline tests/unit_tests/transformer/test_transformer_layer.py:131]: For training would it also make sense to add gradient equivalence tests?

Comment (pengdurice) [reply tests/unit_tests/transformer/test_transformer_layer.py:131]: Thank you for the great suggestion! Just added. Actually, this helped identified a bug in the existing profiling chunking logic in this line 
`bias_output = torch.stack(bias_chunks, dim=0).sum(dim=0) if bias_chunks else None`
It may not manifest for inference, but will impact training since bias_chunks is the chunks of identical bias parameter and we should just take one of them rather than stacking them.

File: megatron/core/transformer/transformer_layer.py [hunk 5]
```diff
@@ -808,7 +822,7 @@ def _forward_mlp(
             # Aggregate chunk outputs
             mlp_output = torch.cat([out for out, _ in outputs], dim=0)
             bias_chunks = [bias for _, bias in outputs if bias is not None]
-            bias_output = torch.stack(bias_chunks, dim=0).sum(dim=0) if bias_chunks else None
+            bias_output = bias_chunks[0] if bias_chunks else None
```
Comment (deepakn94) [inline megatron/core/transformer/transformer_layer.py]: Can you add an assertion before this line checking that they're equal?

Comment (deepakn94) [reply megatron/core/transformer/transformer_layer.py]: Let's add this behind a `if self.config.cuda_graph_impl == 'none'` guard to make sure we don't interfere with CG capture / replay!

Comment (pengdurice) [reply megatron/core/transformer/transformer_layer.py]: I feel adding an assertion is causing additional overheads and actually these biases are the same bias instance (if any) from the same mlp layer. To be on the safe side , I reverted to the original code and do: 
```
if self.config.cuda_graph_impl == 'none': 
     bias_output = bias_chunks[0] if bias_chunks else None
```
Does this make sense? thank you!

Comment (pengdurice) [reply megatron/core/transformer/transformer_layer.py]: @deepakn94 , would you mind take a look at this when you have time? thank you!

File: megatron/core/transformer/transformer_layer.py [hunk 6]
```diff
@@ -809,6 +823,11 @@ def _forward_mlp(
             mlp_output = torch.cat([out for out, _ in outputs], dim=0)
             bias_chunks = [bias for _, bias in outputs if bias is not None]
             bias_output = torch.stack(bias_chunks, dim=0).sum(dim=0) if bias_chunks else None
+            if bias_output is not None:
+                for i in range(1, len(bias_chunks)):
+                    assert torch.allclose(bias_chunks[i], bias_chunks[0])
+            if self.config.cuda_graph_impl == 'none':
+                bias_output = bias_chunks[0] if bias_chunks else None
```
Comment (deepakn94) [inline megatron/core/transformer/transformer_layer.py]: Shouldn't this always be set?

I meant the assertion can be in a `if self.config.cuda_graph_impl == 'none':` block. But on second thought, let's just add a comment saying they should all be identical and remove the assertion.

Comment (pengdurice) [reply megatron/core/transformer/transformer_layer.py]: Hi @deepakn94 thank you for the review, just did that!

Comment (deepakn94) [reply megatron/core/transformer/transformer_layer.py]: Let's get rid of `if self.config.cuda_graph_impl == 'none':` too.

Comment (pengdurice) [reply megatron/core/transformer/transformer_layer.py]: SG! Now it defaults to instead of doing the sum, it uses the first element with a comment, thank you!

Comment (deepakn94) [reply megatron/core/transformer/transformer_layer.py]: Approving from my side.

**Review by deepakn94 — APPROVED** [review]

**Review by santhnm2 — APPROVED** [review]

**Comment by pengdurice** [comment]
Happy Monday! @deepakn94  and @santhnm2 thank you guys so much for reviewing and approving! @Phlip79 would you mind kicking off another CICD test? thank you!

**Comment by pengdurice** [comment]
@santhnm2 , thank you for kicking off the CICD, it still failed approval workflow from more-oncall and core-nemo. Anything that I can do? Do we need to involve additional reviewers? thank you!

**Review by ericharper — APPROVED** [review]

**Comment by ericharper** [comment]
> @santhnm2 , thank you for kicking off the CICD, it still failed approval workflow from more-oncall and core-nemo. Anything that I can do? Do we need to involve additional reviewers? thank you!

Just approved. mcore-oncall is not required and that's just for our internal systems to triage.
